blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp

```diff
@@ -0,0 +1,363 @@
+                const std::size_t nSamplesToConsume = nSamplesToDrop + nSamplesToPublish;
+
+                std::ranges::copy_n(ins[i].begin() + nSamplesToDrop, static_cast<std::ptrdiff_t>(nSamplesToPublish), outs[i].begin());
+                const std::size_t totalDroppedSamples = _nDroppedSamples[i] + nSamplesToDrop;
+
+                publishDroppedSamplesTagIfNotZero(outs[i], totalDroppedSamples);
+                publishInputTags(ins[i], outs[i], nSamplesToDrop, nSamplesToPublish);
+                _nDroppedSamples[i] = 0UZ;
+
+                ins[i].consume(nSamplesToConsume);
+                ins[i].consumeTags(nSamplesToConsume);
+                outs[i].publish(nSamplesToPublish);
+            }
+            _isStreamSynchronized = true;
+        } else {
+            const std::size_t minSamplesBeforeSyncTag = std::ranges::min(ins | std::views::transform([&](const auto& in) { return getNSamplesBeforeSyncTag(in); }));
+            const std::size_t minSamplesOut           = std::ranges::min(outs | std::views::transform([&](const auto& out) { return out.size(); }));
+            const std::size_t nSamplesToCopy          = std::min(minSamplesBeforeSyncTag, minSamplesOut);
+            if (_isStreamSynchronized && nSamplesToCopy > 0UZ) { // all streams are in sync -> write sample before first Sync tag
+                for (std::size_t i = 0; i < nPorts; i++) {
+                    std::ranges::copy_n(ins[i].begin(), static_cast<std::ptrdiff_t>(nSamplesToCopy), outs[i].begin());
+
+                    publishDroppedSamplesTagIfNotZero(outs[i], _nDroppedSamples[i]);
+                    publishInputTags(ins[i], outs[i], 0UZ, nSamplesToCopy);
+
+                    _nDroppedSamples[i] = 0UZ;
+
+                    ins[i].consume(nSamplesToCopy);
+                    ins[i].consumeTags(nSamplesToCopy);
+                    outs[i].publish(nSamplesToCopy);
+                }
+            } else { // streams are NOT in sync -> check back pressure and drop samples if needed
+                for (std::size_t i = 0; i < nPorts; i++) {
+                    const std::size_t nSamplesToDrop = ins[i].size() < static_cast<std::size_t>(max_history_size) ? 0UZ : ins[i].size() - static_cast<std::size_t>(max_history_size);
+                    if (nSamplesToDrop != 0UZ) {
+                        ins[i].consume(nSamplesToDrop);
+                        ins[i].consumeTags(nSamplesToDrop);
+
+                        _nDroppedSamples[i] += nSamplesToDrop;
+                        outs[i].publish(0UZ);
+                        _isStreamSynchronized = false;
+                    }
+                }
+            }
+        }
+        for (std::size_t i = 0; i < nPorts; i++) {
+            if (!ins[i].isConsumeRequested()) {
+                ins[i].consume(0UZ);
+                ins[i].consumeTags(0UZ);
+                outs[i].publish(0UZ);
+            }
+        }
+        return gr::work::Status::OK;
+    }
+
+private:
```

Comment:
Consider removing the `private:` because having the functions below (most could be actually free-standing functions) public makes them easier to unit-test.